mono/metadata/class-init.c

```diff
@@ -773,8 +773,18 @@ mono_class_create_generic_inst (MonoGenericClass *gclass)
 {
 	MonoClass *klass, *gklass;
 
-	if (gclass->cached_class)
+	if (gclass->cached_class) {
+		// The cached_class can sometimes be a GTD in cases like IList
+		// See: https://github.com/mono/mono/issues/7095
+		if (mono_class_is_gtd (gclass->cached_class)) {
+			g_assert (gclass->cached_class->_byval_arg.type != MONO_TYPE_GENERICINST);
+			MonoType *ty = mono_class_gtd_get_canonical_inst (gclass->cached_class);
+			g_assert (ty->type == MONO_TYPE_GENERICINST);
+			MonoGenericClass *gclass2 = ty->data.generic_class;
+			return gclass2->cached_class;
```

Comment:
Do we know `gclass2->cached_class` is always non-NULL here?
What if we did `return mono_class_create_generic_inst (gclass2)` ?

Reply:
Good catch. Pretty sure this pattern is used elsewhere in the codebase though, so I'll have to update that as well.

Reply:
Fwiw I do think it'll always be non-NULL, but can't offer a sufficiently convincing justification of that :)

Reply:
Ah no, the other usage is fine and this is just on me. Fixed.